src/hotspot/share/classfile/placeholders.cpp

```diff
@@ -173,90 +192,41 @@ bool PlaceholderEntry::remove_seen_thread(JavaThread* thread, PlaceholderTable::
-  BasicHashtable<mtClass>::free_entry(entry);
-}
-
-
 // Placeholder objects represent classes currently being loaded.
 // All threads examining the placeholder table must hold the
 // SystemDictionary_lock, so we don't need special precautions
 // on store ordering here.
-PlaceholderEntry* PlaceholderTable::add_entry(unsigned int hash,
-                                              Symbol* class_name, ClassLoaderData* loader_data,
-                                              Symbol* supername){
+PlaceholderEntry* add_entry(Symbol* class_name, ClassLoaderData* loader_data,
+                            Symbol* supername){
   assert_locked_or_safepoint(SystemDictionary_lock);
   assert(class_name != NULL, "adding NULL obj");
 
-  // Both readers and writers are locked so it's safe to just
-  // create the placeholder and insert it in the list without a membar.
-  PlaceholderEntry* entry = new_entry(hash, class_name, loader_data, supername);
-  int index = hash_to_index(hash);
-  Hashtable<Symbol*, mtClass>::add_entry(index, entry);
-  return entry;
+  PlaceholderEntry entry;
+  entry.set_supername(supername);
+  PlaceholderKey key(class_name, loader_data);
+  // Since we're storing this key in the hashtable, we need to increment the refcount.
+  class_name->increment_refcount();
+  bool created;
+  PlaceholderEntry* table_copy = _placeholders.put_if_absent(key, entry, &created);
+  assert(created, "better be absent");
+  return table_copy;
 }
 
-
 // Remove a placeholder object.
-void PlaceholderTable::remove_entry(unsigned int hash,
-                                    Symbol* class_name,
-                                    ClassLoaderData* loader_data) {
-  assert_locked_or_safepoint(SystemDictionary_lock);
-  int index = hash_to_index(hash);
-  PlaceholderEntry** p = bucket_addr(index);
-  while (*p != NULL) {
-    PlaceholderEntry *probe = *p;
-    if (probe->hash() == hash && probe->equals(class_name, loader_data)) {
-      // Delete entry
-      *p = probe->next();
-      free_entry(probe);
-      return;
-    }
-    p = probe->next_addr();
-  }
-}
-
-PlaceholderEntry* PlaceholderTable::get_entry(unsigned int hash,
-                                              Symbol* class_name,
-                                              ClassLoaderData* loader_data) {
+void remove_entry(Symbol* class_name, ClassLoaderData* loader_data) {
   assert_locked_or_safepoint(SystemDictionary_lock);
 
-  int index = hash_to_index(hash);
-  for (PlaceholderEntry *place_probe = bucket(index);
-                         place_probe != NULL;
-                         place_probe = place_probe->next()) {
-    if (place_probe->hash() == hash &&
-        place_probe->equals(class_name, loader_data)) {
-      return place_probe;
-    }
-  }
-  return NULL;
+  PlaceholderKey key(class_name, loader_data);
+  // Decrement refcount in key.
+  class_name->decrement_refcount();
+  _placeholders.remove(key);
```

Comment:
This was in the wrong order.  We should decrement the refcount after removing the entry.